Add documentation and scripts on running e2e test locally (#5489)
kubernetes-sigs/cluster-api-provider-azure: docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md

   - Flavors that leverage internal load balancer are:
     - `apiserver-ilb`
     - `windows-apiserver-ilb`

### Running e2e tests locally using API Server ILB's networking solution

> Building upon the [challenges and solutions](#challenges-and-solutions) from above, running an e2e test locally in a restricted environment calls from some workarounds to the prow templates and the e2e test itself.

Reviewer: ```suggestion
Building upon the [challenges and solutions](#challenges-and-solutions) from above, running an e2e test locally in a restricted environment calls for some workarounds in the prow templates and the e2e test itself.
```

1. We need the apiserver ilb with private endpoints, opinionated CIDRs to the workload cluster's VNet & Subnets, and pre-kubeadm commands updating the etc/hosts of the worker nodes of the workload cluster.

2. Once the template has been modified to be run in local environment using AKS as management cluster, we need to be able to peer the vnets, create private DNS zone for the FQDN of the workload cluster and re-enable blocked NSG ports.

**Note:**

- Following guidance is only for testing, and not a recommendation for any production environment.

- The below steps for self-managed templates only. Does not apply to AKS workload clusters.

> - The below steps are for running the tests in a intuned device. If you are going to run the local tests from a dev machine in Azure, you have to first follow below steps and then proceed ahead.

Author: addressed via 0c7bb63216b299c688bb945af39ee083c4775638

> - The below steps are for running the tests in an intuned device. If you are going to run the local tests from a dev machine in Azure, you have to first follow below steps and then proceed ahead.

Author: Sorry for the wrong wording. The below steps are only required when running the e2e tests from an Dev machine (in Azure).

>   1. Create a managed Identity

Author: addressed via 0c7bb63216b299c688bb945af39ee083c4775638

>   1. Create a user assigned managed identity

Reviewer: ```suggestion
  1. Create a user-assigned managed identity
```

  2. Assign that managed identity a contributor role to your subscription
  3. Set `AZURE_CLIENT_ID_USER_ASSIGNED_IDENTITY`, `AZURE_OBJECT_ID_USER_ASSIGNED_IDENTITY`, `AZURE_USER_ASSIGNED_IDENTITY_RESOURCE_ID` to the user assigned managed identity.

#### Update prow template with apiserver ILB networking solution

There are three sections of a prow template that needs an update.

1. AzureCluster
   - `/spec/networkSpec/apiServerLB`
      - Add FrontendIP

>       - Add an associated private IP to be leveraged by an intern ILB

Reviewer: ```suggestion
      - Add an associated private IP to be leveraged by an internal ILB
```

   - `/spec/networkSpec/vnet/cidrBlocks`
      - Add VNet CIDR
   - `/spec/networkSpec/subnets/0/cidrBlocks`
      - Add Subnet CIDR for the control plane
   - `/spec/networkSpec/subnets/1/cidrBlocks`
      - Add Subnet CIDR for the worker node
2. `KubeadmConfigTemplate` - linux node; Identifiable by `name: .*-md-0`
   - `/spec/template/spec/preKubeadmCommands/0`
      - Add prekubeadm command updating the `/etc/hosts` of the worker node of the type linux.
3. `KubeadmConfigTemplate` - windows node; Identifiable by `name: .*-md-win`
   - `/spec/template/spec/preKubeadmCommands/0`
      - Add prekubeadm command updating the `/etc/hosts` of the worker node of the type windows.

> A Sample kustomize command updating a prow template via its kustomization.yaml is pasted below.

Reviewer 2: ```suggestion
A sample kustomize command for updating a prow template via its kustomization.yaml is pasted below.
```

Author: addressed via 0c7bb63216b299c688bb945af39ee083c4775638

      value: []
    - op: add
      path: /spec/networkSpec/subnets/0/cidrBlocks/-
      value: ${AZURE_CP_SUBNET_CIDR}
- target:
    kind: AzureCluster
  patch: |-
    - op: add
      path: /spec/networkSpec/subnets/1/cidrBlocks
      value: []
    - op: add
      path: /spec/networkSpec/subnets/1/cidrBlocks/-
      value: ${AZURE_NODE_SUBNET_CIDR}
- target:
    kind: KubeadmConfigTemplate
    name: .*-md-0
  patch: |-
    - op: add
      path: /spec/template/spec/preKubeadmCommands
      value: []
    - op: add
      path: /spec/template/spec/preKubeadmCommands/-
      value: echo '${AZURE_INTERNAL_LB_PRIVATE_IP}   ${CLUSTER_NAME}-${APISERVER_LB_DNS_SUFFIX}.${AZURE_LOCATION}.cloudapp.azure.com' >> /etc/hosts
- target:
    kind: KubeadmConfigTemplate
    name: .*-md-win
  patch: |-
    - op: add
      path: /spec/template/spec/preKubeadmCommands/-
      value:
        powershell -Command "Add-Content -Path 'C:\\Windows\\System32\\drivers\\etc\\hosts' -Value '${AZURE_INTERNAL_LB_PRIVATE_IP} ${CLUSTER_NAME}-${APISERVER_LB_DNS_SUFFIX}.${AZURE_LOCATION}.cloudapp.azure.com'"
```

#### Peer Vnets of the management cluster and the workload cluster

> Peering VNets, creation of private DNS zone with the FQDN of the workload cluster and updating of NSGs of management and workload cluster can be achieved by `scripts/peer-vnets.sh`

Reviewer: ```suggestion
Peering VNets, creating a private DNS zone with the FQDN of the workload cluster, and updating NSGs of the management and workload clusters can be achieved by running `scripts/peer-vnets.sh`.
```

> Peering VNets, creation of private DNS zone with the FQDN of the workload cluster and updating of NSGs of management and workload cluster can be achieved by `peer-vnets.sh` present in `scripts/peer-vnets.sh`

Reviewer 2: ```suggestion
Peering VNets, creation of private DNS zone with the FQDN of the workload cluster and updating of NSGs of management and workload cluster can be achieved by `scripts/peer-vnets.sh`
```

Author: addressed via 0c7bb63216b299c688bb945af39ee083c4775638
